wfsim/core.py

```diff
@@ -612,91 +607,147 @@ def __init__(self, config):
-            if len(truth_buffer):
-                self.get_truth(instruction, truth_buffer)
 
-            # The instruction index is saved for chunking
-            self.instruction_index = ix
+            # D) Run all clusters before the current source
+            stop_at_this_group = False
+            for ibqs in instb_queue:
+                for ptype in [1, 2, 3, 4]:
+                    mask = instb_type[ibqs] == ptype
+                    if np.sum(mask) == 0: continue # No such instruction type
+                    instb_run = instb_indx[ibqs[mask]] # Take hold of todo list
+
+                    if self.symtype(ptype) in ['s1', 's2']:
+                        stop_at_this_group = True # Stop group iteration
+                        _instb_run = np.array_split(instb_run, len(instb_run))
+                    else: _instb_run = [instb_run] # Small trick to make truth small
+
+                    # Run pulse simulation for real
+                    for instb_run in _instb_run:
+                        for instb_secondary in self.sim_data(instb[instb_run]):
+                            ib = np.where(~instb_filled)[0][:len(instb_secondary)]
+                            instb[ib] = instb_secondary
+                            instb_filled[ib] = True
+
+                        if len(truth_buffer): # Extract truth info
+                            self.get_truth(instb[instb_run], truth_buffer)
+
+                        instb_filled[instb_run] = False # Free buffer AFTER copyting into truth buffer
+
+                if stop_at_this_group: break
+                self.digitize_pulse_cache() # from pulse cache to raw data
+                yield from self.ZLE()
+                
+            self.source_finished = len(inst_queue) == 0 and np.sum(instb_filled) == 0
+        pbar.close()
 
-        self.source_finished = True
-        self.digitize_pulse_cache()
-        yield from self.ZLE()
-        
     @staticmethod
     def symtype(ptype):
-        return ['s1', 's2'][ptype - 1]
+        return ['s1', 's2', 'unknown', 'pi_el', 'pmt_ap'][ptype - 1]
 
     def sim_data(self, instruction):
-        ptype = self.symtype(instruction['type'])
+        ptype = self.symtype(instruction['type'][0])
         self.pulses[ptype](instruction)
-        self.pulses['ele_ap'](self.pulses[ptype])
         self.pulses['pmt_ap'](self.pulses[ptype])
-        #
-        for pt in [ptype, 'ele_ap', 'pmt_ap']:
-            self.pulses_cache += getattr(self.pulses[pt], '_pulses')
-        #
-        if len(self.pulses['ele_ap']._pulses)  > 0:
-            self.pulses['pmt_ap'](self.pulses['ele_ap'])
-            self.pulses_cache += getattr(self.pulses['pmt_ap'], '_pulses')
-
-        if len(self.pulses_cache) > 0:
-            self.last_pulse_end_time = max(self.last_pulse_end_time,
-                np.max([p['right'] for p in self.pulses_cache]) * 10)
-
+        
+        for pt in [ptype, 'pmt_ap']:
+            _pulses = getattr(self.pulses[ptype], '_pulses')
```

Comment:
Should be 'pt' as the key, the double-counting is coming from here.

Reply:
Good catch! I did some fast simulations and it looks promising! 
![S1a_for_pr](https://user-images.githubusercontent.com/27280678/68970754-4f273c80-07e8-11ea-9f18-656d7e657775.png)
![S2a_for_pr](https://user-images.githubusercontent.com/27280678/68970763-52bac380-07e8-11ea-9d8c-1e7082a9021d.png)
Only the S1_b seems to be still off.... (was the same before the redesign of the chunk mechanism)
![S1b_for_pr](https://user-images.githubusercontent.com/27280678/68970798-68c88400-07e8-11ea-8fae-01984d02b645.png)